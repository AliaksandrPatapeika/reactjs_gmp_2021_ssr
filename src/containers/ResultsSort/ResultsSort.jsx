import './ResultsSort.less';

import {assign, filter, head} from 'lodash';
import {useRouter} from 'next/router';
import React from 'react';
import {useDispatch, useSelector} from 'react-redux';

import {setSortBy, setSortOrder} from '../../actions/movies';
import Button from '../../components/Button';
import Select from '../../components/Select';
import SortIcon from '../../components/SortIcon';
import {SortBy} from '../../constants';
import {getQueryString} from '../../utils';

const ResultsSort = () => {
  const sortBy = useSelector((state) => state.query.sortBy);
  const sortOrder = useSelector((state) => state.query.sortOrder);
  const query = useSelector((state) => state.query);
  const dispatch = useDispatch();
  const router = useRouter();
  const {id} = router.query;

  const getSortValue = () => head(
    filter(SortBy, (option) => option.value === sortBy)
  );

  const setSortByHandler = (selectItem) => {
    /* for the sorting when the film details page is open */
    if (id) {
      dispatch(setSortBy(selectItem.value));
    } else {
      router.push({
        pathname: '/search',
        search: getQueryString(assign(query, {sortBy: selectItem.value}))
      });
    }
  };

  const toggleSortOrder = () => (sortOrder === 'desc' ? 'asc' : 'desc');

  const setSortOrderHandler = () => {
    /* for the sorting order when the film details page is open */
    if (id) {
      dispatch(setSortOrder(toggleSortOrder()));
    } else {
      router.push({
        pathname: '/search',
        search: getQueryString(assign(query, {sortOrder: toggleSortOrder()}))
      });
    }
  };

  return (
    <div className="resultsSortContainer">
      <div className="sortBy">SORT BY</div>
      <Select
        className="select resultSortSelect"
        classNamePrefix="select"
        id="resultSort"
        options={SortBy}
        onChange={setSortByHandler}
        value={getSortValue()}
      />
      <Button className="arrow" onClick={setSortOrderHandler}>
        <SortIcon sortOrder={sortOrder} />
      </Button>
    </div>
  );
};

export default ResultsSort;
