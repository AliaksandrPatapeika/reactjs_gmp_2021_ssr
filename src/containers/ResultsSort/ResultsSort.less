@import "../../less/variables";

.resultsSortContainer {
	display: flex;
	padding-bottom: 17px;
	align-items: center;

	.sortBy {
		color: @gray;
		font-size: 20px;
		font-family: @NetflixSans-Regular;
		white-space: nowrap;
	}

	.arrow {
		cursor: pointer;
		margin-left: 10px;
		color: @red;
		font-size: 22px;
		background-color: transparent;

		&:hover,
		&:focus {
			color: @red-hover;
		}
	}

	.resultSortSelect {
		width: 210px;
		margin-left: 10px;

		.select__control {
			background-color: transparent;

			.select__single-value {
				color: @white;
				font-size: 22px;
				font-family: @TextaAlt-Regular;
			}

			.select__dropdown-indicator {
				font-size: 22px;
			}
		}
	}
}
