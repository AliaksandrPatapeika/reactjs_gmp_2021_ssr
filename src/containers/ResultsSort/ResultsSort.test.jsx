import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';

import ResultsSort from './ResultsSort';

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: {id: '12'}
  })
}));

describe('ResultsSort', () => {
  test('renders correctly', () => {
    const mockStore = configureMockStore();
    const initialState = {
      query: {
        filter: '',
        sortBy: 'release_date',
        sortOrder: 'desc'
      }
    };
    const store = mockStore(initialState);
    const resultsSort = renderer
      .create(
        <Provider store={store}>
          <ResultsSort />
        </Provider>
      )
      .toJSON();

    expect(resultsSort)
      .toMatchSnapshot();
  });
});
