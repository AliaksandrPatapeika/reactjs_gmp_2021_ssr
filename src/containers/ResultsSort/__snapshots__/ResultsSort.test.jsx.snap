// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsSort renders correctly 1`] = `
<div
  className="resultsSortContainer"
>
  <div
    className="sortBy"
  >
    SORT BY
  </div>
  <div
    className="select resultSortSelect css-2b097c-container"
    id="resultSort"
    onKeyDown={[Function]}
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="select__control css-yk16xz-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="select__value-container select__value-container--has-value css-g1d714-ValueContainer"
      >
        <div
          className="select__single-value css-1uccc91-singleValue"
        >
          RELEASE DATE
        </div>
        <div
          className="css-b8ldur-Input"
        >
          <div
            className="select__input"
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-autocomplete="list"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              disabled={false}
              id="react-select-resultSort-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "boxSizing": "content-box",
                  "color": "inherit",
                  "fontSize": "inherit",
                  "label": "input",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "1px",
                }
              }
              tabIndex="0"
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": 0,
                  "left": 0,
                  "overflow": "scroll",
                  "position": "absolute",
                  "top": 0,
                  "visibility": "hidden",
                  "whiteSpace": "pre",
                }
              }
            >
              
            </div>
          </div>
        </div>
      </div>
      <div
        className="select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className="select__indicator-separator css-1okebmr-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <i
            aria-hidden="true"
            className="fa fa-caret-down"
          />
        </div>
      </div>
    </div>
  </div>
  <button
    className="arrow"
    disabled={false}
    id={null}
    onClick={[Function]}
    type="button"
  >
    
    <i
      aria-hidden="true"
      className="fa fa-sort-amount-desc"
    />
  </button>
</div>
`;
